src/sinks/gcp/stackdriver_metrics.rs

```diff
@@ -92,162 +96,325 @@ fn default_endpoint() -> String {
+    };
+
+    let unit = value_type.as_str_name().to_string();
+
+    googleapis_proto::TimeSeries {
+        metric: Some(googleapis_proto::Metric {
+            r#type: format!(
+                "custom.googleapis.com/{}/metrics/{}",
+                data_point.namespace, data_point.name
+            ),
+            labels: data_point.tags.unwrap_or_else(Default::default),
+        }),
+        resource: Some(googleapis_proto::MonitoredResource {
+            r#type: config.resource.r#type.clone(),
+            labels: config.resource.labels.clone(),
+        }),
+        metadata: None,
+        metric_kind: metric_kind.into(),
+        value_type: value_type.into(),
+        points: vec![googleapis_proto::Point {
+            interval: Some(googleapis_proto::TimeInterval {
+                end_time: Some(timestamp_proto(end_time)),
+                start_time: Some(timestamp_proto(start_time)),
+            }),
+
+            value: Some(googleapis_proto::TypedValue { value: Some(value) }),
+        }],
+        unit,
     }
 }
 
+struct GrpcEventSink {
+    config: StackdriverConfig,
+}
+
 #[async_trait::async_trait]
-impl HttpSink for HttpEventSink {
-    type Input = Metric;
-    type Output = Vec<Metric>;
-    type Encoder = StackdriverMetricsEncoder;
+impl StreamSink<EventArray> for GrpcEventSink {
+    async fn run(self: Box<Self>, mut input: stream::BoxStream<'_, EventArray>) -> Result<(), ()> {
+        let auth = match self.config.auth.build(Scope::MonitoringWrite).await {
+            Ok(auth) => auth,
+            Err(e) => {
+                error!(message = "Fatal gcp_stackdriver_metrics error.", %e);
+                return Err(());
+            }
+        };
 
-    fn build_encoder(&self) -> Self::Encoder {
-        StackdriverMetricsEncoder
-    }
+        let uri = if let Ok(uri) = self.config.endpoint.as_str().parse::<http::Uri>() {
+            uri
+        } else {
+            error!(message = "Error when parsing uri.", %self.config.endpoint);
+            return Err(());
+        };
+
+        let channel = Channel::builder(uri)
+            .tls_config(tonic::transport::ClientTlsConfig::new())
+            .map_err(|e| {
+                error!(message = "Fatal gcp_stackdriver_metrics error.", %e);
+            })?
+            .connect_lazy();
+
+        let mut stopwatch = tokio::time::interval(Duration::from_secs(1));
+        let mut aggregate = HashMap::<String, MetricDataPoint>::with_capacity(GCP_MAX_BATCH_SIZE);
+        let mut time_tracker = StartTimeTracker::new();
+        let mut last_time = Instant::now();
+        let mut buffer = Vec::<MetricDataPoint>::new();
+        let mut client =
+            MetricServiceClient::<Channel>::with_interceptor(channel, interceptor(auth));
+
+        loop {
+            if aggregate.len() == GCP_MAX_BATCH_SIZE {
+                while last_time.elapsed() < GCP_SMALLEST_CREATE_TIMESERIES_CALL_FREQUENCY {
```

Comment:
Could we have a comment here explaining why this is necessary.

Reply:
Done